### PR TITLE
Add cpu-affinity

### PR DESCRIPTION
**Pull request by k0ekk0ek**
This PR adds:
* Support for binding a server to a specific CPU and
* Support for binding an IP address only to selected servers (or child processes)

I still need to write test cases. Actually, I still need to test :wink:, but I'm very interested in initial feedback regarding the options and their configuration.

### CODE REVIEW

File: bitset.c
Comment (wcawijngaards) [inline bitset.c:10]: This file needs the license.  Also include config.h at the start, for portability defines, such as for "inline" or malloc.

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: Either this should be >=, or the asserts in bitset_set and unset should not be >=.  Are you sure you are or-ing the final bits in the bitstring.

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: None of the functions in this header file have a documentation, but I was surprised to find that the bitset_or function is dest = dest|A|B  and not dest=A|B.  (This in the manner of, surprises are not good to have in code, as a general principle.  This isn't really anything wrong with the code, per se).

File: configlexer.lex
Comment (wcawijngaards) [inline configlexer.lex]: This one can be on one line, like `tls-port` above. Also for the `xfrd-cpu-affinity` definition.  If you like.

**Review by wcawijngaards — COMMENTED** [review]
Nice code, this is a partial review.  The config looks nice, not sure about ip-address additions, for not changing the existing stuff for people that do not use the feature, it could be its own line to associate parameters with an ip-address or is that lots of code? So that some spelling error does not put a cpu afficinity on someones IPv6 address, even if they are not using that feature.

File: configparser.y
Comment (wcawijngaards) [inline configparser.y]: The callers of `parse_number` do not use or need the errno. The errno wipe and restore seem unused here, remove it?

File: bitset.c
Comment (k0ekk0ek) [reply bitset.c:10]: Fixed

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: You're absolutely right. The `nsd_bitset_or` suffered from more brainfarts. Fixed now.

File: configlexer.lex
Comment (k0ekk0ek) [reply configlexer.lex]: Done

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: Done

**Comment by k0ekk0ek** [comment]
Added printing of server process affinity on startup at info level so that users can see which server process listens on a given interface. `*` is used for all, `-` for none, `x`, `x y` or `x-y` is used for individual servers or server-ranges.

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: is this `srcset2->size` instead of `srcset1`? Looks like a mistake to me, because `i` then indexes `srcset2->bits`.

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: You don't always have a spare byte with partial bits in it, so >= is not always the correct comparison, right? So >. And sometimes, if size%8!=0 there is another byte?

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: Not >=, right? Because the array is malloc(n), not malloc(n+1), right?

File: bitset.h
Comment (wcawijngaards) [inline bitset.h]: You need to make sure the 1<< is unsigned int, not unsigned char, that is the type of mask.  Otherwise you'd run out of bitspace for the shift.  It may be nice to make that explicit, with a cast.

File: cpuset.c
Comment (wcawijngaards) [inline cpuset.c]: s/cpuset.h/cpuset.c/

File: cpuset.c
Comment (wcawijngaards) [inline cpuset.c]: need config.h at the start of the includes because of defines for different include standards in config.h.

Comment (wcawijngaards) [reply cpuset.c]: If you do this, the _GNU_SOURCE define may then already be defined in config.h, and then you have to check with ifdef for that before you define it for sched.h above.

File: cpuset.h
Comment (wcawijngaards) [inline cpuset.h]: You include this in nsd.h, but there the _GNU_SOURCE define is not there, and thus the sched.h include below does not (always) work.

Comment (wcawijngaards) [reply cpuset.h]: If you made the nsd_cpuset_type opaque, and the define kept in the cpuset.c sourcefile we can contain the platform-weirdness to that file for it.  Without exposing it to other headfiles and all other c files through nsd.h.

**Review by wcawijngaards — COMMENTED** [review]
partial commit of comments.

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: That's a mistake. Should indeed read `srcset2`.

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: Correct

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: Correct

File: bitset.h
Comment (k0ekk0ek) [reply bitset.h]: Correct

File: nsd.c
Comment (wcawijngaards) [inline nsd.c]: This change is not part of this PR.  Also, why? Was there a user request for it?

File: nsd.c
Comment (wcawijngaards) [inline nsd.c:1122]: As a question, why do you need to check for this? Couldn't they just be disjunct?

File: nsd.conf.5.in
Comment (wcawijngaards) [inline nsd.conf.5.in]: delete this line, you have a spurious '-a' in the man page.

Comment (wcawijngaards) [reply nsd.conf.5.in]: If what you wanted was a new paragraph (which seems possible from the whitespace in the nroff source layout), this is with `.IP` on its own on a line to separate the paragraphs.

File: nsd.conf.sample.in
Comment (wcawijngaards) [inline nsd.conf.sample.in]: N-1 for this example, 0..N-1 is the range of CPU numbers.

File: server.c
Comment (wcawijngaards) [inline server.c:2791]: You should not remove this memset, perhaps move it inside the `add_tcp_handler` function, but you cannot delete it, it would make a fixed bug reappear.

**Review by wcawijngaards — APPROVED** [review]
Appreciate your code as always.  Some more items for consideration, the remainder is unit tests, so I'm okay with the PR.

File: nsd.conf.sample.in
Comment (k0ekk0ek) [reply nsd.conf.sample.in]: The two numbers can be the same, but they don't have to. It was meant to indicate to the user that he/she can specify a cpu per server process, but I guess you could read that as well. Maybe I should use something like `X` or something on that line.

File: nsd.c
Comment (k0ekk0ek) [reply nsd.c]: It was part of a patch with the initial version of cpu-affinity. I thought I'd include it. It seems to make sense that this be treated as a warning as opposed to an error, but I'll happily remove it if you think that's not the case(?)

File: nsd.c
Comment (k0ekk0ek) [reply nsd.c:1122]: Yes, they can be disjunct. There's a couple of reasons why it's like this, but they're not strong technical reasons. We can change things around if desired. In no particular order:
 * Other operating systems don't offer the same kind of functionality, or at least, not in the same way. Most that offer the user to specify cpu affinity do not allow cpus to be in multiple sets. I wanted to keep things the same for all platforms as much as possible. With that in mind I want to allow the user to be more specific on both Linux and FreeBSD, but still require him/her to define the total set.
 * Privileges. The 'root` user often starts the daemon, afterwhich the process drops permissions to the specified user. I haven't tried every combination in great detail yet, but I know at least a couple of other operating systems allow to bind to a specific set only when called with root privileges. I require the user to specify the total set in the configuration, hoping that lower privileged processes can choose one cpu from the specific set later on. I reasoned that if assigned to specific cpus, lower privileged processes would certainly not be allowed to pick something outside the set. This is not at all a given though.
 * The configuration only has options for `xfrd` and specific `server` processes, this one actually specifies what cpus the `main` process will bind to. My reasoning was: main runs on x-z, it's children pick a specific one from that set.
 * The prototype with cpu-affinity implemented for NSD3 had it.

All-in-all it felt like a sane way to implement it, and thus mainly comes down to taste, but as stated before, we can certainly change things around if desired.

File: nsd.c
Comment (wcawijngaards) [reply nsd.c]: It depends on whether the customer wants that as error in the logs or warning. So we'd need to ask some to know. In the past I got the sentiment that, the pidfile is stale means that NSD had crashed earlier, a strong error condition.

Comment (k0ekk0ek) [reply nsd.c]: I'll remove this for now. It's not that important for these changes

Comment (wcawijngaards) [reply nsd.c]: Sure, we should check with users to see if they desire this change, at some point.

File: nsd.c
Comment (wcawijngaards) [reply nsd.c:1122]: okay, thank you for the explanation.

File: tpkg/bad_cpu_affinity.tdir/bad_cpu_affinity.pre
Comment (wcawijngaards) [inline tpkg/bad_cpu_affinity.tdir/bad_cpu_affinity.pre:13]: This value does not exist for me.  On freeBSD it seems to be from `sysctl -n hw.ncpu` that prints something like `4` or `sysctl: cannot stat /proc/sys/hw/ncpu: No such file or directory` on Linux.  On Linux, /proc/cpuinfo contains lines of text, and I think you need `grep "^processor" /proc/cpuinfo  | wc -l` that also prints the plain number.

Comment (k0ekk0ek) [reply tpkg/bad_cpu_affinity.tdir/bad_cpu_affinity.pre:13]: Ok. Good to know it doesn't exist everywhere. I just grabbed something that existed on my Fedora machine. Indeed, the `grep` version might work on more systems, I'll use that.

Comment (wcawijngaards) [reply tpkg/bad_cpu_affinity.tdir/bad_cpu_affinity.pre:13]: Sorry about that, nproc works fine for me too.  And I had not read the common.sh part yet where you have the shell subroutine for that.  So you can just use that.

File: configparser.y
Comment (k0ekk0ek) [reply configparser.y]: Done

File: cpuset.c
Comment (k0ekk0ek) [reply cpuset.c]: Fixed

File: cpuset.c
Comment (k0ekk0ek) [reply cpuset.c]: Done

File: nsd.c
Comment (k0ekk0ek) [reply nsd.c]: Removed

File: server.c
Comment (k0ekk0ek) [reply server.c:2791]: Restored `memset`

File: nsd.conf.sample.in
Comment (k0ekk0ek) [reply nsd.conf.sample.in]: Swapped out `N` for an actual number

File: nsd.conf.5.in
Comment (k0ekk0ek) [reply nsd.conf.5.in]: Removed the `-a`

File: cpuset.h
Comment (k0ekk0ek) [reply cpuset.h]: Turns out `_GNU_SOURCE` is defined in `config.h`. I removed it from the header. On second thought though, I think `cpuset` is better moved to compat(?)

Comment (wcawijngaards) [reply cpuset.h]: Ok. but not in compat, this is not a libc feature or something, but a wrapper, right? I mean it gets compiled every time?

Comment (k0ekk0ek) [reply cpuset.h]: Correct. Both FreeBSD and Linux have the same macro's though. My plan would be to update `configure.ac` and simply introduce an extra typedef for Linux (or FreeBSD) (Linux uses cpu_set_t, FreeBSD uses cpuset_t) and get rid of the wrapper functions. It's unlikely that any other platform will get support for the feature as most others don't even support it in the first place. With that in mind: `setproctitle` is also not a libc feature, should that be removed from `compat`?

Comment (wcawijngaards) [reply cpuset.h]: Sure, if you want that.  Setproctitle is on freebsd, I thought, hence that made sense, because it is a replace_func c file?

File: tpkg/cutest/cutest_bitset.c
Comment (wcawijngaards) [inline tpkg/cutest/cutest_bitset.c]: wrong indentation. s/  //.

File: tpkg/do-tests
Comment (wcawijngaards) [inline tpkg/do-tests]: do you mean `echo $NEED_CPU_AFFINITY` that would select the elements of that list for skip.

**Review by wcawijngaards — APPROVED** [review]
Completed review, no comments.  This is nice code :-)

File: cpuset.h
Comment (k0ekk0ek) [reply cpuset.h]: The cpuset got moved to `compat` and the set is now allocated dynamically and worked on based on the `cpuset_*` interfaces in NetBSD, which (in theory), means that NetBSD should be supported as well.

File: tpkg/do-tests
Comment (k0ekk0ek) [reply tpkg/do-tests]: Indeed

File: tpkg/cutest/cutest_bitset.c
Comment (k0ekk0ek) [reply tpkg/cutest/cutest_bitset.c]: Corrected

**Comment by k0ekk0ek** [comment]
Processed the comments, cleaned up the commits and ran some basic tests. We'd still have to verify the cpu-affinity features are not compiled-in on unsupported architectures and we' to run all of the tests, but that maybe something we want to do as part of the release process(?)

**Comment by k0ekk0ek** [comment]
For some reason the `bitset_or_even_odd` and `bitset_or_even_even` are a hit and miss, I'll look into that right away.

**Comment by k0ekk0ek** [comment]
Should be good to go now. @wcawijngaards you want me to test on all platforms before merging or should I go ahead and merge?

**Comment by wcawijngaards** [comment]
I am happy with your commit either way, and a good code branch is what we need.  So what you think is better; test then merge sounds good, but merge then test does not look wrong as well.

**Comment by k0ekk0ek** [comment]
Updated the release notes and did a quick test run. Going to try compilation on some additional platforms and will merge if that works

**Comment by wcawijngaards** [comment]
Okay!

**Comment by k0ekk0ek** [comment]
So, I compiled and tested on FreeBSD and OpenBSD (and Linux of course). Merging this PR.

**Pull request merged**
